### DISCUSSION TITLE
Does the new UnsafeAccessor not work with protected visibility?

### DISCUSSION THREAD

**Discussion by ScarletKuro** [dotnet/runtime#108950] [Q&A]
Hi,

I wanted to access the following method from the `ComponentBase` class:

```C#
protected Task DispatchExceptionAsync(Exception exception) => _renderHandle.DispatchExceptionAsync(exception);
```

I initially used the following `UnsafeAccessor`:

```C#
[UnsafeAccessor(UnsafeAccessorKind.Method, Name = "DispatchExceptionAsync")]
private static extern Task DispatchExceptionAsync(ComponentBase component, Exception exception);
...
await DispatchExceptionAsync(component, ex);
```

However, I encountered an exception saying that the method `DispatchExceptionAsync` could not be found. As a workaround, I had to access the private `_renderHandle` directly:

```C#
[UnsafeAccessor(UnsafeAccessorKind.Field, Name = "_renderHandle")]
private static extern ref RenderHandle RenderHandle(ComponentBase component);
...
await RenderHandle(component).DispatchExceptionAsync(ex);
```

I did not find any limitations mentioned in:

- [UnsafeAccessorAttribute documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafeaccessorattribute?view=net-8.0)
- [GitHub documentation on UnsafeAccessors](https://github.com/dotnet/runtime/blob/main/docs/design/features/unsafeaccessors.md)

I expected it to work with any visibility, as long as the signature is defined correctly.

**Comment by KeterSCP** [comment]
Can you share a complete version of your code (and also which .NET version you are using)? I've tried running this sample and it works as expected (both on .NET 8 and .NET 9):

```csharp
using System.Runtime.CompilerServices;
using Microsoft.AspNetCore.Components;

var c = new MyComponent();
var ex = new Exception("Test");

await DispatchExceptionAsync(c, ex);

[UnsafeAccessor(UnsafeAccessorKind.Method, Name = "DispatchExceptionAsync")]
static extern Task DispatchExceptionAsync(ComponentBase component, Exception exception);

class MyComponent : ComponentBase
{
}
```

Version of `Microsoft.AspNetCore.Components` nuget is `8.0.10`

**Reply by ScarletKuro** [reply]
> Can you share a complete version of your code

https://github.com/MudBlazor/MudBlazor/pull/10043

> which .NET version you are using

8.0.306 SDK

**Reply by ScarletKuro** [reply]
Thanks to you I discovered that this doesn't work on just WASM. So it's definitely a bug then.

**Reply by ScarletKuro** [reply]
Tried 8.0.403, doesn't work either.

**Reply by ScarletKuro** [reply]
Issue created: https://github.com/dotnet/runtime/issues/108957
Repro: https://github.com/user-attachments/files/17406667/BlazorApp2Bug.zip
